orttraining/orttraining/core/framework/torch/torch_proxy.cc

```diff
@@ -150,15 +151,29 @@ void InvokeRunner(
     // from Pytorch.
     PyObject* py_obj = PyTuple_GetItem(result_ptr.get(), 0);
     if (is_training_mode) {
-      const auto& refcnt = Py_REFCNT(py_obj);
-      // We don't need do ref increase here because, python returns tensor.grad_fn as part of
-      // tuple, who increased the refcnt already (and tensor persist until the backward kernels completed).
-      // Pytorch also increases refcnt before apply() return, so we should expect refcount >= 2.
-      // We say "at least" 2 because user could increase the context refcnt as well in their autograd forward()
-      // and backward() functions.
-      ORT_ENFORCE(refcnt >= 2, "Ref count of context should be 2, but actually it's ", refcnt, ".");
-      if (refcnt > 2) {
-        LOGS_DEFAULT(WARNING) << "Autograd context refcnt > 2";
+      if (py_obj == Py_None) {
+        {
+          static std::once_flag once;
+          std::call_once(once, [] {
+            LOGS_DEFAULT(WARNING) << "Under training mode, autograd context found to be Py_None.";
+          });
+        }
+      } else {
+        const auto refcnt = Py_REFCNT(py_obj);
+        // We don't need do ref increase here because, python returns tensor.grad_fn as part of
+        // tuple, who increased the refcnt already (and tensor persist until the backward kernels completed).
+        // Pytorch also increases refcnt before apply() return, so we should expect refcount >= 2.
+        // We say "at least" 2 because user could increase the context refcnt as well in their autograd forward()
+        // and backward() functions.
+        ORT_ENFORCE(refcnt >= 2, "Ref count of context should be 2, but actually it's ", refcnt, ".");
+        if (refcnt > 2) {
+          {
+            static std::once_flag once;
+            std::call_once(once, [] {
+              LOGS_DEFAULT(WARNING) << "Autograd context refcnt > 2, refcnt: " << refcnt;
```

Comment:
Should this be a warning considering the comment above?

Reply:
From the observation, some PythonOp are exported with an attribute training model = True. While during forward, none of the inputs of the PythonOp requires grad, as a result, context will be None. So this looks like to be an acceptable case, previously we printed warning for every step, which is not friendly to users. Maybe printing a warning once is enough in case that can be clues for us to investigate any issues.

Reply:
Oh maybe you mean, it can be verbose or just skipped? I would prefer to leave some minor warning here for knowing what happens.

Reply:
Think twice, let's move them into verbose .